[hailtop] Added gather and modified AsyncWorkerPool

jigold opened the pull request:
FYI: @cseed

danking commented on hail/python/hailtop/utils/utils.py:
```diff
@@ -26,34 +26,93 @@ def async_to_blocking(coro):
+        self._futures = []
+
+        self._results = [None] * len(coros)
+        self._error = None
```
This should be a list and we should append to it. As written, one worker could set the error and trigger the done event, another worker could then overwrite the error with a new error and trigger the done event again, and then finally the wait thread wakes up and raises the second error. We can just always through the first exception, but I think it will be less confusion if we know it's always the first exception that gets raised.

Longer term, we might need to write a new exception that contains all the other exceptions. I'm not exactly sure how to do that, so let's punt on that.

jigold commented on hail/python/hailtop/utils/utils.py:
```diff
@@ -26,34 +26,93 @@ def async_to_blocking(coro):
-        async with self._sem:
             try:
-                await f(*args, **kwargs)
+                res = await coro
```
I think my way is clearer. The behavior of this function should insert the exception as the result for that coroutine rather than raising an exception if raise_exceptions is False.

jigold commented on hail/python/hailtop/utils/utils.py:
```diff
@@ -26,34 +26,93 @@ def async_to_blocking(coro):
+        self.count = len(coros)
+        self.n_finished = 0
+
+        self._queue = asyncio.Queue(maxsize=1000)
```
I don't think we can do that if we want to return immediately after the first exception. I think we'd want to do that?

danking replied:
ok

jigold commented on hail/python/hailtop/utils/utils.py:
```diff
@@ -26,34 +26,93 @@ def async_to_blocking(coro):
-    def __init__(self, parallelism):
-        self._sem = asyncio.Semaphore(parallelism)
-        self._count = 0
+async def gather(*coros, parallelism=10, return_exceptions=False):
```
I swapped the names, but the logic is still the same. `raise_exceptions=True` rather than `return_excpetions=False`.

jigold commented on hail/python/hailtop/batch_client/aioclient.py:
```diff
@@ -358,19 +357,21 @@ def create_job(self, image, command, env=None, mount_docker_socket=False,
+            coros.append(self._submit_job(batch.id, specs))
 
-        await self.pool.wait()
+        await gather(*coros, parallelism=2)
```
cool

jigold commented:
This code doesn't quite work in the case of cancelling. I'm working on a fix.

jigold commented:
I fixed the cancel -- you have to take partial functions as the argument if you want to cancel when the first exception has been thrown rather than waiting for all coroutines to finish before throwing the exception. I reverted raise_exceptions back to return_exceptions because I like that better and it matches `asyncio.gather()`.
